### ISSUE TITLE
Add support for Pod Security Admission

### ISSUE THREAD

**Issue by nwmac** [rancher/dashboard#7119]
Documentation: https://kubernetes.io/docs/concepts/security/pod-security-admission/
Epic: https://github.com/rancher/rancher/issues/38701

# Issues

Grouped features
- https://github.com/rancher/dashboard/issues/7617
- https://github.com/rancher/dashboard/issues/7677
- https://github.com/rancher/dashboard/issues/7628
- https://github.com/rancher/dashboard/issues/7575
- https://github.com/rancher/dashboard/issues/7875
- https://github.com/rancher/dashboard/issues/7999

Smaller UI iterations
- https://github.com/rancher/dashboard/issues/7597
- https://github.com/rancher/dashboard/issues/7583
- https://github.com/rancher/dashboard/issues/7584
- ~https://github.com/rancher/dashboard/issues/7684~ Done in PR

# Implementations

## 1. Namespace creation/edit UI

Create new tab 'Pod Security Admission' to be shown before 'Labels & annotations' that allows user to configure the PSA for the namespace.

This would have 3 checkboxes to enable enforce, audit, warn - for each when checked, dropdown to chose policy `privileged`, `baseline`, or `restricted` and an input box to enter a version (defaults to text 'latest').

The UI will take this form and add/update the labels required for these settings.

On the labels and annotations tab, if there are PSA labels, we will by default hide these but add a button/link to allow the user to see them in the labels list - see the mock for labels further down. Note the mock shows the system labels being readonly, but I think we should allow them to be edited. 

UX design for tab contents:

![image](https://user-images.githubusercontent.com/1955897/199688475-abd0c19a-52e1-4bb7-81d6-b49f0a55f387.png)

Note: The defaults for the select box should be 'Baseline' - the mock up above shows the three possibilities.

## 2. Project creation/edit UI

As above - user can use the same UI but for Project create/edit. The backend will take care of sync'ing the labels from the project to the namespaces in that project

## 3. Project/Namespace list

When a namespace has PSA labels, we want to show an icon (lock) next to the namespace name - on hover shows the PSA config for that namespace - e.g. "Enforce Privileged (version), Audit Baseline (latest)" (as a vertical list of the up to 3 possibilities)

## 4. Namespace Detail View

In the masthead for the resource, for the labels, we change the display of the labels that relate to PSAs - those starting with `pod-security.kubernetes.io/` - instead we present these as: "ICON Enforce Privileged (version)" - on hover user sees the labels that we have hidden and replaced with this alternate display - note this can be 2 labels - one for the config and one for the version.

## 5. Workload view (e.g. detail view for Pod, Deployment etc)

Present the error message better when the error is due to a PSA violation.

Instead of:

![image](https://user-images.githubusercontent.com/1955897/198034445-f207655b-f902-47b8-ae5e-4bdb4f109218.png)

we use a regex on the message to detect a Pod PSA violation message and show:

"ICON Friendly message (TBD) ..." 

Hovering on the "..." would show the original full error message.

## 6. Workload edit screen

After creation of a workload/pod, check response for the Warning header (indicated a PSA violation when mode is set to warn) and present a growl in a similar fashion to the error above, without hover).

Note: This requires backend changes to get access to the warning header - https://github.com/rancher/rancher/issues/39413

## 7. Cluster Management - Cluster Admission Configurations

Add a new resource view in Cluster Management that lists the Admission Configurations - there will be a new Custom Resource that mirrors the PSA one to be used only for cluster creation.

Add a create/edit screen that allows the user to configuration an admission configuration (see Admission Configuration here https://kubernetes.io/docs/tutorials/security/cluster-level-pss_

## 8. Cluster Creation RKE1 and RKE2

When creating an RKE1 or RKE2 cluster, the user will be able to select an existing Admission Configuration that will provided the default PSS configuration for the cluster.

![image](https://user-images.githubusercontent.com/1955897/203307615-cb9fbea3-fa46-4513-9e24-8076714349b4.png)

**Comment by cnotv** [comment]
> PSPs are going away in 1.25. PSAs are the new thing. As I understand it they are a lot simpler than PSPSs. If you look at that document, for namespaces, you can add a label that configures the mode for the pods in a given namespace, so I imagine we'll want to add that to the detail view for a namespace and also the edit view.
You can also declare exemptions for pods, so that will require a change to the pods detail/edit screens.
I don't know if there is a log from the admission controller that we could show to help users understand what pods were blocked by the configuration.
Rancher does not officially support k8s 1.25, so one task is to bring up a 1.25 cluster and import it into Rancher - we'll need that for dev/testing.

**Comment by kwwii** [comment]
This shows the creation ui pattern as well as the changes necessary to the Labels and Annotations
![PSA-creation](https://user-images.githubusercontent.com/617368/198114028-7a68f48a-f400-4e56-8183-ddc9b84471c3.png)

**Comment by kwwii** [comment]
This shows the creation ui pattern and the changes necessary to the key/value ui pattern

![PSA-creation](https://user-images.githubusercontent.com/617368/198116328-684491ab-4a75-4abe-8eae-de1bdbf28f45.png)

**Comment by kwwii** [comment]
Here is an example of how the alerts could look. This uses the "pod_security" icon from eos-icons.com (the SUSE icon set).

![RestrictedMessage](https://user-images.githubusercontent.com/617368/198122377-15360687-8733-4421-90ed-546fd1bbd892.png)

**Comment by kwwii** [comment]
https://xd.adobe.com/view/7d2d1b4d-4d4f-420c-b2d8-2271839a4727-287e/

**Comment by cnotv** [comment]
Some notes as reminder:
- Exceptions are displayed only on pod creation as separated resource within the Cluster Manager
- Icon in the error/warning needs to be bound to the existing map and extend the banner component
- Border for the Labels tab content should be exceptional for this case and not global
- Username in exceptions have no autocomplete for this iteration
- Namespace have no autocomplete as are not created at this stage yet

**Comment by gaktive** [comment]
Related backend work on k8s 1.25 support, including the proposed design for the CRD and appropriate fields: https://github.com/rancher/rancher/issues/38701

**Comment by cnotv** [comment]
> `/shell/plugins/steve/hybrid-class.js` `get labels()` is used by the Labels component and has the code that omits some of them. for the state of an existing resource, `/shell/plugins/dashboard-store/resource-class.js` everything that revolves around `stateObj` , in particular `stateDisplay/stateDescription, stateIcon, stateBackground, stateColor`, etc

**Comment by kwwii** [comment]
The tab in the UI should be named "Pod Security Standards", the titles on the page should be "Pod Security Configuration" (instead of the current "Pod Security Admission" in the mockup) and "Exceptions"

**Comment by cnotv** [comment]
@kwwii should I add the [SUSE icon](https://eos-icons.com/?iconName=pod_security&type=static) to the [Rancher icon](https://github.com/rancher/icons)?
I can see that we are adding the resource PSA but in the design we finally have 3. Are these going to vary in length based on the resource? 
In the table how are these going to be recognized?

@nwmac I have been creating a `pod.security.admission` as list, edit and model, name TBD. In the edit I do not have the title, in the list is not showing at all due some error for the missing model and in the model I have no idea which one to extend: Norman, Steve, Workload, etc. I have no idea when to use which one.
We should add all these as [documentation for the resources](https://rancher.github.io/dashboard/guide/customising-how-k8s-resources-are-presented), possibly in a more practical approach.
Another case would be also how to customize the table cell, but probably this should have an example in the [styleguide](https://rancher.github.io/storybook) since it's a component.

**Comment by cnotv** [comment]
> The tab in the UI should be named "Pod Security Standards", the titles on the page should be "Pod Security Configuration" (instead of the current "Pod Security Admission" in the mockup) and "Exceptions"

Why do we have inconsistent names?
Also how is going to be called in the sidenav of the cluster manager?

**Comment by kwwii** [comment]
The names are intended to reflect the naming in the k8s docs. As I understand it, we have "standards" which are defined by a "configuration" which can include "exceptions". The "admission" is used in connection with the controller, afaict

**Comment by kwwii** [comment]
@cnotv I don't quite understand your question about the icons. Can you explain?

**Comment by cnotv** [comment]
In order to have the icon, I must add it to the other repository, is that correct? Should I make a separated PR request or what?

**Comment by kwwii** [comment]
@cnotv yes, you will need to add the icon to the icon repo (as I understand it at least)

**Comment by cnotv** [comment]
Ok, here's the PR https://github.com/rancher/icons/pull/44

**Comment by cnotv** [comment]
@kwwii should we not maybe use this toggler for the show/hide labels?
https://rancher.github.io/storybook/?path=/story/components-form-toggleswitch--toggle-switch

![Screenshot 2022-11-07 at 15 48 22](https://user-images.githubusercontent.com/5009481/200339331-c382536e-8bd2-4737-97d2-119eeffdb4d2.png)

**Comment by kwwii** [comment]
@cnotv Yes, that it is a good idea

**Comment by cnotv** [comment]
Please refer to this issue for technical specifications and requirements: https://github.com/rancher/dashboard/issues/7176#issuecomment-1284718619
FYI @nwmac @kwwii

**Comment by cnotv** [comment]
@jiaqiluo What will happens to the current labels and annotations which we have everywhere?
In the current design the annotations are removed, is that matching the initial requirement or I am mixing things up?

**Comment by jiaqiluo** [comment]
> @jiaqiluo What will happens to the current labels and annotations which we have everywhere? In the current design the annotations are removed, is that matching the initial requirement or I am mixing things up?

hi @cnotv , do you mean the labels annotations currently used for PSP?  It is expected to see those removed from the objects by the backend on 1.25 clusters. 
Please correct me if I misunderstand your question.

**Comment by jiaqiluo** [comment]
@cnotv Oh, I see what you mean now. We should still support setting the labels and annotations, they are separate features unrelated to the PSA.

> In the current design the annotations are removed,

I never hear about this, it might be some miscommunication somewhere, but UI should keep the existing support for labels and annotations.

**Comment by cnotv** [comment]
@jiaqiluo ok, so does it mean that this new label is going to be displayed only for Namespaces and Projects editors? Should we keep the other `Label and Annotation`?

<img width="891" alt="Screenshot 2022-11-21 at 22 27 28" src="https://user-images.githubusercontent.com/5009481/203161867-38b9db29-04f9-4aaf-93e9-6bbf130ab582.png">


If also this information is wrong, could you please define where should this view be added, on top of the resource itself?

**Comment by cnotv** [comment]
Summary of the clarification:
- PSA are going to be added to Namespace editor tab and optionally to Projects, if ready
- The labels will display the PSA values and be protected, but values can be hardcoded, [based on specification](https://kubernetes.io/docs/concepts/security/pod-security-admission/#pod-security-admission-labels-for-namespaces)
- The Namespace detail, if contains a label of type PSA, it should be mapped in a nice way
- Workload creation should return a warning on save and should be displayed in a growl
- PSA should be added as new Rancher resource and listed in the Cluster Management, allowing users to create templates
- Inside the Cluster editor 
  - If Kubernetes version is >=1.25 it should hide the current PSP dropdown and show the PSA dropdown
  - The PSA template will be displayed with the list of resources by name, default value empty
  - The dropdown values needs to be changed on version change with the default values

Also based on the current development, the Label component should not be used as it applies everywhere.

**Comment by cnotv** [comment]
@kwwii I have a case where if we type values which make the field disable and impossible to remove, the user will be stuck with that. What's your idea about it to prevent it?

**Comment by kwwii** [comment]
Namespace PSA are easily removable by deselecting the checkbox. The values should not persist if the user has deselected the checkbox.

Cluster PSAs are NOT removable once the cluster has been provisioned.

If the user enters enters these under labels and not as a PSA I would only expect the UI to show them when the user enters them, and if the "show/hide" is set to show. This means that if the user has the system labels set to hidden, and enters a PSA as label instead of as a PSA, the UI would still hide these labels, only showing them when the user selected "Show system labels".

**Comment by cnotv** [comment]
This is what I mean:

https://user-images.githubusercontent.com/5009481/204484530-2fc9ccf3-e244-4644-b446-50917b5023c2.mov

**Comment by cnotv** [comment]
This is 

https://user-images.githubusercontent.com/5009481/204515565-23ddd3c0-c64c-49a8-bff1-4b5337df4c3e.mov

with the integrated view:

**Comment by cnotv** [comment]
@nwmac when we add labels, we use `value.setLabels($event)`, we'll have to deal with something similar also for the PSA.

**Comment by kwwii** [comment]
I think we need to leave everything editable, only toggling the view of the labels we have defined as system labels. 

We might also think about adding something to help avoid confusion by the user if they enter a system label in the normal label ui - we could automatically toggle the viewing of system labels to true

**Comment by kwwii** [comment]
Run the setting of inactive/active when you load the tab content. That would allow a user to enter a system label and edit it….until the page refreshes or the user changes tabs and back, etc.

**Comment by cnotv** [comment]
In that case we would have the same scenario, where the user saves something in a wrong way and be unable to remove it.

I would suggest at this point to stick to the design and then eventually we may go through all the points directly at the end or even with the QA. It's kinda time consuming at this state to check all the points for edge cases.

**Comment by cnotv** [comment]
New issue created for RK1 https://github.com/rancher/dashboard/issues/7575

**Comment by cnotv** [comment]
In RKE2 Cluster editing, the logic about k8s version needs to be updated in order to keep also existing options. I give the assumption that nothing needs to be changed.

https://user-images.githubusercontent.com/5009481/204583828-0394aaab-7253-4cab-9657-b72623255804.mov

**Comment by cnotv** [comment]
Splitting up in more issues as this one keeps growing and we already invested a too much time.

**Comment by cnotv** [comment]
@nwmac I think that we missed a very key point here for the issue. The labels are alway a pair of values, e.g.:
```
    pod-security.kubernetes.io/enforce: <value>
    pod-security.kubernetes.io/enforce-version: <value> 
```

This would require to refactor every component and logic affected. Most of the time we simply go through the resource (`value`) and consider the labels as single values.
This will include refactoring of all the mentioned cases, e.g. https://github.com/rancher/dashboard/issues/7597

**Comment by nwmac** [comment]
@cnotv If this is for the masthead to hide the PSA labels, I'd suggest we add to the model another getter like the details one, which if present would return the labels to display - it can then filter out the PSA ones

**Comment by cnotv** [comment]
We have already the whole object inside the view, the tab and also the component, so we just need a function to filter it.

**Comment by cnotv** [comment]
As mentioned, points 1 to 4 are going to be tackled with https://github.com/rancher/dashboard/issues/7617 as every feature can already be added.

**Comment by cnotv** [comment]
> Here is an example of how the alerts could look. This uses the "pod_security" icon from the SUSE icon set (https://eos-icons.com/?iconName=pod_security&type=static).
> 
> ![RestrictedMessage](https://user-images.githubusercontent.com/617368/198134283-c266f154-f17b-46b0-92fd-6bd6e387ecd4.png)

There's something missing from the requirement, not sure if is removed, missing or misunderstood.
We have this type of error which is stated to be displayed on over:

![image](https://user-images.githubusercontent.com/1955897/198034445-f207655b-f902-47b8-ae5e-4bdb4f109218.png)

Should we add a togler to extend the banner, a tooltip, an icon to toggle the text, or what's the plan with the rest of the warning?

**Comment by kwwii** [comment]
I don't think that the rest of the message has any value to the user so it is ok to hide it. Someone feel free to correct me if I am wrong.

**Comment by nwmac** [comment]
I agree @kwwii

**Comment by cnotv** [comment]
We still have to split the errors in more points looks like. I can see there's more errors chained, like `Deployment does not have minimum availability`

**Comment by cnotv** [comment]
I noticed that we are lacking of documentation on how to add a new types to the navigation, so I created a new  issue.
https://github.com/rancher/dashboard/issues/7637

**Comment by cnotv** [comment]
New update: Labels will be updated everywhere altogether with Annotations which are then kept for the Namespace.

**Comment by cnotv** [comment]
@kwwii @richard-cox 
As we have 22 occurrences for the component `Labels`, I am not going to change the title and spacing with this implementation, unless you strictly want to update it everywhere. It will require `:show-header="false"` to be applied everywhere.
22 pages to be checked on QA are quite too much.

<img width="283" alt="Screenshot 2022-12-16 at 14 22 16" src="https://user-images.githubusercontent.com/5009481/208107622-a0314176-2470-4662-aec8-9be5137660c0.png">

**Comment by gaktive** [comment]
We need a similar warning for k8s 1.23 & .24 users, though more along the lines of guidance of the retirement of PSPs. https://github.com/rancher/dashboard/issues/7791 filed

**Comment by cnotv** [comment]
Removed issue related to the table as erroneously created for this scope: we have no errors in the list view.

**Comment by cnotv** [comment]
Based on @jiaqiluo comment, `defaultPodSecurityAdmissionConfigurationTemplateId` is going to be replaced by `defaultPodSecurityAdmissionConfigurationTemplateName`

**Comment by cnotv** [comment]
Closing issue as everything is completed and tested.
